### PR TITLE
Scaling down fees

### LINKED ISSUE #230: Add a task to rescale fees on chain (down by 10-12)
Since some fees are hard coded in Plug we need a way to scale down all fees in CENNZnet.

### PR DESCRIPTION
**Pull request by KarishmaBothara**
Rescale fees on chain to match GA balances down by 10-12
Closes #230 

Tested

### CODE REVIEW

**Comment by aliXsed** [comment]
My first comment is, your PR should have a better title and some useful description.

**Comment by aliXsed** [comment]
Second, a PR should be linked to a ticket number. As here the code base is public, you'd better first create a public issue, saying what you are going to do and then if this PR addresses that issue in the description you'd mention "closes #...".

File: crml/transaction-payment/src/lib.rs [hunk 1]
```diff
@@ -88,7 +88,7 @@ pub trait Trait: frame_system::Trait {
 	/// The fee to be paid for making a transaction; the per-byte portion.
 	type TransactionByteFee: Get<BalanceOf<Self>>;
 
-	/// Convert a weight value into a deductible fee based on the currency type.
+	/// Convert a weight value into a deductible fee based on the currency type and scale it.
```
Comment (jordy25519) [inline crml/transaction-payment/src/lib.rs]: ` and scale it` part is implementation specific, not necessary for the type

File: crml/transaction-payment/src/lib.rs [hunk 2]
```diff
@@ -487,7 +487,14 @@ mod tests {
 	pub struct WeightToFee(u64);
 	impl Convert<Weight, u64> for WeightToFee {
 		fn convert(t: Weight) -> u64 {
-			WEIGHT_TO_FEE.with(|v| *v.borrow() * (t as u64))
+			let _linear_weight = WEIGHT_TO_FEE.with(|v| *v.borrow() * (t as u64));
```
Comment (aliXsed) [inline crml/transaction-payment/src/lib.rs]: As you have used this variable, its name shouldn't start with _

File: crml/transaction-payment/src/lib.rs [hunk 3]
```diff
@@ -487,7 +487,14 @@ mod tests {
...
+			let _fee_divider: u64 = 1_000_000_000_000;
```
Comment (aliXsed) [inline crml/transaction-payment/src/lib.rs]: The name shouldn't start with _

Comment (aliXsed) [inline crml/transaction-payment/src/lib.rs]: const FEE_DIVIDER: u64 = 1..

File: crml/transaction-payment/src/lib.rs [hunk 4]
```diff
@@ -688,7 +695,7 @@ mod tests {
 		let len = ext.len() as u32;
 		ExtBuilder::default()
 			.base_fee(5)
-			.weight_fee(2)
+			.weight_fee(10_000_000_000_000)
```
Comment (aliXsed) [inline crml/transaction-payment/src/lib.rs]: Why did you need to change a test? Couldn't you just add a new test?

File: runtime/src/impls.rs [hunk 5]
```diff
@@ -81,14 +81,22 @@ impl Convert<u128, Balance> for CurrencyToVoteHandler {
...
-impl<C: Get<Balance>> Convert<Weight, Balance> for LinearWeightToFee<C> {
+impl<C: Get<Balance>> Convert<Weight, Balance> for ScaleLinearWeightToFee<C> {
 	fn convert(w: Weight) -> Balance {
 		// cennznet-node a weight of 10_000 (smallest non-zero weight) to be mapped to 10^7 units of
 		// fees, hence:
 		let coefficient = C::get();
-		Balance::from(w).saturating_mul(coefficient)
+		let _linear_weight = Balance::from(w).saturating_mul(coefficient);
+		// Scale down fees 10^-12
+		let _fee_divider: Balance = 1_000_000_000_000;
```
Comment (aliXsed) [inline runtime/src/impls.rs]: FEE_DIVIDER should be defined in one place and used wherever needed.

File: runtime/src/impls.rs [hunk 6]
```diff
@@ -81,14 +81,22 @@ impl Convert<u128, Balance> for CurrencyToVoteHandler {
...
+pub struct ScaleLinearWeightToFee<C>(sp_std::marker::PhantomData<C>);
 
-impl<C: Get<Balance>> Convert<Weight, Balance> for LinearWeightToFee<C> {
+impl<C: Get<Balance>> Convert<Weight, Balance> for ScaleLinearWeightToFee<C> {
 	fn convert(w: Weight) -> Balance {
 		// cennznet-node a weight of 10_000 (smallest non-zero weight) to be mapped to 10^7 units of
 		// fees, hence:
 		let coefficient = C::get();
-		Balance::from(w).saturating_mul(coefficient)
+		let _linear_weight = Balance::from(w).saturating_mul(coefficient);
```
Comment (aliXsed) [inline runtime/src/impls.rs]: _linear.. --> linear..

File: runtime/tests/tests.rs [hunk 7]
```diff
@@ -551,6 +552,22 @@ fn generic_asset_transfer_works_without_fee_exchange() {
 	});
 }
 
+#[test]
+fn compute_fee_scaling() {
+	ExtBuilder::default().build().execute_with(|| {
+		let dispatch_info = DispatchInfo {
+			weight: 1000000000,
```
Comment (jordy25519) [inline runtime/tests/tests.rs]: style: can `_` separate the numbers for readability

File: runtime/src/impls.rs [hunk 8]
```diff
@@ -81,14 +81,22 @@ impl Convert<u128, Balance> for CurrencyToVoteHandler {
...
-impl<C: Get<Balance>> Convert<Weight, Balance> for LinearWeightToFee<C> {
+impl<C: Get<Balance>> Convert<Weight, Balance> for ScaleLinearWeightToFee<C> {
 	fn convert(w: Weight) -> Balance {
 		// cennznet-node a weight of 10_000 (smallest non-zero weight) to be mapped to 10^7 units of
 		// fees, hence:
 		let coefficient = C::get();
-		Balance::from(w).saturating_mul(coefficient)
+		let linear_weight = Balance::from(w).saturating_mul(coefficient);
+		// Scale down fees 10^-12
+		let fee_divider: Balance = 1_000_000_000_000;
```
Comment (aliXsed) [inline runtime/src/impls.rs]: I have added a parameter called ScaleDownFactor to Runtime for this purpose. If I merge first, you can reuse that.

File: runtime/src/impls.rs [hunk 9]
```diff
@@ -81,14 +81,22 @@ impl Convert<u128, Balance> for CurrencyToVoteHandler {
...
+		let scaled_weight = linear_weight.checked_div(fee_divider).unwrap_or(linear_weight);
+		if scaled_weight.is_zero() {
+			linear_weight
+		} else {
+			scaled_weight
```
Comment (aliXsed) [inline runtime/src/impls.rs:104]: This is probably not a good idea. Let me give you an example. Suppose the scale down factor is 7. If the fee is 8 the result will be 1, If the fee is 6 the result will be 6. It means the fees now have lost their proportions.

**Review by aliXsed — CHANGES_REQUESTED** [review]

File: runtime/src/impls.rs (hunk 9 above)
Comment (KarishmaBothara) [reply runtime/src/impls.rs:104]: I get your point.. what if I have min_weight..
if scaled_weight.is_zero() && linear_weight < min_weight {
  linear_weight
} else if scaled_weight.is_zero() {
 min_weight
} else {
 scaled_weight
}

File: runtime/src/impls.rs [hunk 10]
```diff
@@ -81,14 +81,27 @@ impl Convert<u128, Balance> for CurrencyToVoteHandler {
...
 		// fees, hence:
 		let coefficient = C::get();
-		Balance::from(w).saturating_mul(coefficient)
+		let linear_weight = Balance::from(w).saturating_mul(coefficient);
+		// Scale down fees 10^-12
+		let fee_divider: Balance = 1_000_000_000_000;
+		let min_weight: Balance = 10_000_000;
+		let scaled_weight = linear_weight.checked_div(fee_divider).unwrap_or(linear_weight);
+		if scaled_weight.is_zero() {
+			if linear_weight < min_weight {
```
Comment (jordy25519) [inline runtime/src/impls.rs:98]: Now that this is quite branchy it's worth testing the cases

**Review by jordy25519 — APPROVED** [review]
lgtm! pending some unit test cases

File: runtime/src/impls.rs [hunk 11]
```diff
@@ -81,14 +81,27 @@ impl Convert<u128, Balance> for CurrencyToVoteHandler {
...
+pub struct ScaleLinearWeightToFee<C>(sp_std::marker::PhantomData<C>);
 
-impl<C: Get<Balance>> Convert<Weight, Balance> for LinearWeightToFee<C> {
+impl<C: Get<Balance>> Convert<Weight, Balance> for ScaleLinearWeightToFee<C> {
 	fn convert(w: Weight) -> Balance {
 		// cennznet-node a weight of 10_000 (smallest non-zero weight) to be mapped to 10^7 units of
 		// fees, hence:
 		let coefficient = C::get();
-		Balance::from(w).saturating_mul(coefficient)
+		let linear_weight = Balance::from(w).saturating_mul(coefficient);
```
Comment (jordy25519) [inline runtime/src/impls.rs:91]: AFAIK sometimes `w` could be `0` in that case, this function should return `0` (+test also)

Comment (KarishmaBothara) [reply runtime/src/impls.rs:91]: sure.. will add a test for this.

File: runtime/src/impls.rs (hunk 8 above)
Comment (KarishmaBothara) [reply runtime/src/impls.rs]: done

Comment (aliXsed) [reply runtime/src/impls.rs]: Now after "**use**"ing the ScaleDownFactor here, you can do `let fee_divider = ScaleDownFactor::get();`

**Review by aliXsed — APPROVED** [review]
